### PR TITLE
Place all invocations of read on stream objects into try-catch blocks

### PR DESCRIPTION
**Pull request by xiaofan-cerebras**
All invocations of the read function on stream objects can throw "insufficient data" exceptions. They should be placed inside try-catch blocks in order to achieve maximum possible parsing depth.

### CODE REVIEW

**Comment by ebertland** [comment]
Which case(s) actually caused an uncaught exception or situation where `malformed(true)` was not called?

**Comment by xiaofan-cerebras** [comment]
The original problem happened to a packet with a corrupted IPv4 "total length" field. The IP packet carries an ICMP payload of the type "TIMESTAMP_REPLY". This type of ICMP requires an additional 8 bytes of data after the ICMP header, and this part of the parsing was outside the try-catch block, leading to an uncaught exception.

However, after reviewing the code, I feel that this kind of uncaught exceptions can happen in other places where the read function is called outside of try-catch blocks, such as IPv4 options parsing.

**Comment by xiaofan-cerebras** [comment]
I've pushed a second commit to roll back the per-header try-catch blocks and use a child class of InputMemeoryStream class to handle the insufficient_data exception.

**Comment by xiaofan-cerebras** [comment]
The `read` function of the InputMemoryStream class is a template function. Since C++ doesn't allow virtual template functions, I chose to target the `can_read` function instead.

File: include/tins/memory_helpers.h [hunk 1]
```diff
@@ -113,9 +114,17 @@ class InputMemoryStream {
         return TINS_LIKELY(size_ >= byte_count);
     }
 
-    void read(void* output_buffer, size_t output_buffer_size) {
-        if (!can_read(output_buffer_size)) {
+    virtual bool check_can_read(size_t byte_count) const {
+        if (!can_read(byte_count)) {
             throw insufficient_data();
+            return false;
```
Comment (ebertland) [inline include/tins/memory_helpers.h]: This return is unreachable. Did you add it because of a warning or other check?

File: include/tins/memory_helpers.h [hunk 2]
```diff
@@ -141,11 +150,40 @@ class InputMemoryStream {
     void read(HWAddress<6>& address);
     void read(IPv4Address& address);
     void read(IPv6Address& address);
-private:
+protected:
```
Comment (ebertland) [inline include/tins/memory_helpers.h]: I don't see either of these data members used in the PduInputMemoryStream derived class.

File: src/icmp.cpp [hunk 3]
```diff
@@ -78,6 +73,7 @@ ICMP::ICMP(const uint8_t* buffer, uint32_t total_sz)
     }
     // Attempt to parse ICMP extensions
     try_parse_extensions(stream);
+
```
Comment (ebertland) [inline src/icmp.cpp]: nit: Unnecessary blank line.

File: src/ip.cpp [hunk 4]
```diff
@@ -130,6 +125,7 @@ IP::IP(const uint8_t* buffer, uint32_t total_sz) {
             options_.push_back(option(opt_type));
         }
     }
+
```
Comment (ebertland) [inline src/ip.cpp]: nit: Unnecessary blank line.

(The author doesn't use much whitespace (or any?) within functions. I'm not a big fan of that style, but we should try to follow his style.)

File: src/tcp.cpp [hunk 5]
```diff
@@ -125,6 +120,7 @@ TCP::TCP(const uint8_t* buffer, uint32_t total_sz) {
             stream.skip(len);
         }
     }
+
```
Comment (ebertland) [inline src/tcp.cpp]: nit: Unnecessary whitespace.

**Review by ebertland — CHANGES_REQUESTED** [review]
Looks good. Just a few minor changes requested.

**Comment by ebertland** [comment]
Did you run the test suite?

**Comment by ebertland** [comment]
There are a few places where you change the function argument type to from InputMemoryStream to PduInputMemoryStream where it's probably not necessary. In other words, just allow implicit cast to the base class. Since the functions you overrode are virtual, it should work fine. The only reason to change that would be to reduce the size of the change if we ever try to upstream this enhancement. I'm happy with the way you did it, though.

File: include/tins/memory_helpers.h (hunk 1 above)
Comment (xiaofan-cerebras) [reply include/tins/memory_helpers.h]: I was worried about a compiler warning for not returning on the conditional, but apparently the compiler is happy with not having the return here, so I've pushed a change to remove it.

File: include/tins/memory_helpers.h (hunk 2 above)
Comment (xiaofan-cerebras) [reply include/tins/memory_helpers.h]: I originally initialized these two members in the child class constructor, which requires that these members be protected instead of private. I've resolved this issue by calling the parent class constructor in the child class constructor.

**Comment by xiaofan-cerebras** [comment]
> There are a few places where you change the function argument type to from InputMemoryStream to PduInputMemoryStream where it's probably not necessary. In other words, just allow implicit cast to the base class. Since the functions you overrode are virtual, it should work fine. The only reason to change that would be to reduce the size of the change if we ever try to upstream this enhancement. I'm happy with the way you did it, though.

I think it's a good idea to use the parent class reference instead of the child class one, this minimizes the code changes and gives the callers of the functions more flexibility.

I've pushed a commit to make this change.

**Comment by xiaofan-cerebras** [comment]
> Did you run the test suite?

I reran the initial failed simulation using this build of the library, and the simulation is able to pass. Running the simulation with the normal libtins build reproduced the failure. So these changes seem to fix the problem.

**Comment by ebertland** [comment]
> > Did you run the test suite?
> 
> I reran the initial failed simulation using this build of the library, and the simulation is able to pass. Running the simulation with the normal libtins build reproduced the failure. So these changes seem to fix the problem.

I mean the unit tests that come with libtins. See `Running tests` in README.md.

**Comment by xiaofan-cerebras** [comment]
> > > Did you run the test suite?
> > 
> > 
> > I reran the initial failed simulation using this build of the library, and the simulation is able to pass. Running the simulation with the normal libtins build reproduced the failure. So these changes seem to fix the problem.
> 
> I mean the unit tests that come with libtins. See `Running tests` in README.md.

I have run the unit tests with the changes. All tests passed.

**Review by ebertland — APPROVED** [review]

**Comment by ebertland** [comment]
Looks good. You should do a "squash and merge" for this change.
